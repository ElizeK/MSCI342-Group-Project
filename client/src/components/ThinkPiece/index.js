import React, { Component, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import CssBaseline from "@material-ui/core/CssBaseline";
import { MuiThemeProvider, createTheme } from "@material-ui/core/styles";
import { Typography } from "@material-ui/core";
import history from "../Navigation/history";
import "@fontsource/oswald";
import "@fontsource/inter";
import { Grid, Toolbar, Button, Paper, FormControl, InputLabel, Select, MenuItem, TextField, Box } from '@mui/material';
import { LastPageOutlined } from '@material-ui/icons';

// This theme sets the background color for when you scroll behind the screen
const theme = createTheme({
    palette: {
        type: 'dark',
        background: {
            default: "#1b1b1b"
        },
        primary: {
            main: "#FFFFFF",
        },
        secondary: {
            main: "#1b1b1b",
        },
    },
});

const opacityValue = 1;

const useStyles = makeStyles((theme) => ({
    body: {
        backgroundColor: "#1b1b1b",
        overflow: "hidden",
        color: "#1b1b1b"
    },
    backgroundColor: {
        backgroundColor: "#1b1b1b"
    },
    paper: {
        overflow: "hidden",
    },
    message: {
        opacity: opacityValue,
        maxWidth: 250,
        paddingBottom: theme.spacing(2),
    },
    formControl: {
        margin: theme.spacing(1),
        minWidth: 200,
    },
    selectEmpty: {
        marginTop: theme.spacing(2),
    },
    bullet: {
        display: 'inline-block',
        margin: '0 2px',
        transform: 'scale(0.8)',
    },
    textField: {
        "& .MuiInputBase-root": {
            color: 'white'
        },
        "& .MuiOutlinedInput-root": {
            "& > fieldset": { borderColor: "white" },
        },
        "& .MuiInputLabel-root": { color: 'white' },
        "& .MuiOutlinedInputLabel-root": { color: 'white' },
        "& .MuiOutlinedInput-root:hover": {
            "& > fieldset": {
                borderColor: "white"
            }
        },
        "& .MuiOutlinedInput-root.Mui-focused": {
            "& > fieldset": {
                borderColor: "white"
            }
        }
    },
    select: {
        '&:before': {
            borderColor: 'white',
        },
        '&:after': {
            borderColor: 'white',
        },
        '&:not(.Mui-disabled):hover::before': {
            borderColor: 'white',
        },
        '& .MuiOutlinedInput-notchedOutline': {
            border: "1px solid white !important"
        },
        '& .MuiOutlinedInput-notchedOutline.Mui-focused': {
            borderColor: "white !important"
        },
        '& .MuiSvgIcon-root': {
            fill: "white !important",
        }
    },
    heading: {
        color: "white",
        fontFamily: "Oswald",
        fontSize: 55
    },
    subHeading: {
        color: "white",
        fontFamily: "Inter",
        fontSize: 22,
        fontStyle: "normal",
        fontWeight: 300

    },
    navbarItem: {
        fontFamily: 'Inter',
        fontStyle: "normal",
        fontWeight: 400,
        fontSize: 16,
        cursor: "pointer"
    },
    inputLabelRoot: {
        color: "white",
    }
}));

const ButtonAppBar = () => {
    const classes = useStyles();
    return (
        <div>
            <Toolbar>
                <Typography style={{ marginRight: 10 }}></Typography>
                <Button
                    color="inherit"
                    onClick={() => history.push('/Landing')}
                >
                    <Typography className={classes.navbarItem} variant="h6" component="div" sx={{ flexGrow: 1 }}>
                        Pulse News
                    </Typography>
                </Button>
                <Typography style={{ marginRight: 50 }}></Typography>
                <Button
                    color="inherit"
                    onClick={() => history.push('/')
                    }
                >
                    <Typography className={classes.navbarItem}>Home</Typography>
                </Button>
                <Typography style={{ marginRight: 50 }}></Typography>
                <Button
                    color="inherit"
                    style={{ cursor: "pointer" }}
                    onClick={() => history.push('/Search')}
                >
                    <Typography className={classes.navbarItem}>Search</Typography>
                </Button>
                <Typography style={{ marginRight: 50 }}></Typography>
                <Button
                    color="inherit"
                    style={{ cursor: "pointer" }}
                    onClick={() => history.push('/ThinkPiece')}
                >
                    <Typography className={classes.navbarItem}>Thinkpiece</Typography>
                </Button>
                <Typography style={{ marginRight: 50 }}></Typography>
                <Button
                    color="inherit"
                    style={{ cursor: "pointer" }}
                    onClick={() => history.push('/Profile')}
                >
                    <Typography className={classes.navbarItem}>Profile</Typography>
                </Button>

            </Toolbar>
        </div>

    );
}


const ThinkPiece = () => {
    const Categories = {
        BUSINESS: "Business",
        ENTERTAINMENT: "Entertainment",
        GENERAL: "General",
        HEALTH: "Health",
        SCIENCE: "Science",
        SPORTS: "Sports",
        TECHNOLOGY: "Technology"
    }

    const [topic, setTopic] = useState(Categories.GENERAL);
    const [title, setTitle] = useState("");
    const [summary, setSummary] = useState("");
    const [url, setUrl] = useState("");
    const [content, setContent] = useState("");

    const classes = useStyles();

    const handleSubmit = (e) => {
        e.preventDefault()
        submitThinkpiece()
    }

    const submitThinkpiece = async () => {
        const url = "/api/thinkpieces";

        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                userId: 1,
                title: title,
                content: content,
                summary: summary,
                topic: topic,
                url: url
            })
        });

        const body = await response.json();

        if (response.status !== 200) throw Error(body.message);
        //set back to original states when submitted 
        setTopic(Categories.GENERAL);
        setContent("");
        setSummary("");
        setUrl("");
        setTitle("");

        return body;
    }

    return (

        <div data-testid="think-piece-component">
            <ButtonAppBar
                backgroundColor="secondary"
            ></ButtonAppBar>
            <Grid
                container
                direction="row"
                // justifyContent="center"
                style={{ minHeight: '100vh' }}
                className={classes.backgroundColor}
            >
                <Grid item style={{ marginTop: 50 }} xs={5}>
                    <Box ml={7} p={2}>
                        <Typography className={classes.heading}>Think Piece</Typography>
                        <Typography className={classes.subHeading}>Write your own piece based on news you've read</Typography>
                    </Box>
                    <Box ml={7} p={2}>
                        <TextField
                            required
                            fullWidth
                            id="outlined-basic"
                            label="Title"
                            value={title}
                            variant="outlined"
                            className={classes.textField}
                            onChange={(e) => setTitle(e.target.value)}
                            InputLabelProps={{
                                style: { color: '#fff' },
                            }}
                            data-testid='Title'
                        />
                    </Box>
                    <Box ml={7} p={2}>
                        <TextField
                            required
                            fullWidth
                            id="outlined-multiline-required"
                            label="Summary"
                            multiline
                            maxRows={4}
                            value={summary}
                            className={classes.textField}
                            onChange={(e) => setSummary(e.target.value)}
                            InputLabelProps={{
                                style: { color: '#fff' },
                            }}
                            data-testid='Summary'
                        />
                    </Box>
                    <Box ml={7} p={2}>
                        <FormControl fullWidth>
                            <InputLabel style={{ color: "#fff" }}>Topic</InputLabel>
                            <Select
                                value={topic}
                                label="Topic"
                                onChange={(e) => setTopic(e.target.value)}
                                className={classes.select}
                                style={{ color: "#fff" }}
                                required
                                data-testid='Topic'

                            >
                                {
                                    Object.keys(Categories).map(key => Categories[key]).map((categoryName) =>
                                        <MenuItem value={categoryName}>{categoryName}</MenuItem>
                                    )
                                }
                            </Select>
                        </FormControl>
                    </Box>
                    <Box ml={7} p={2}>
                        <TextField
                            required
                            fullWidth
                            id="outlined-required"
                            label="Article URL"
                            value={url}
                            className={classes.textField}
                            onChange={(e) => setUrl(e.target.value)}
                            InputLabelProps={{
                                style: { color: '#fff' },
                            }}
                            data-testid='URL'
                        />
                    </Box>
                </Grid>
                <Grid item xs={1}></Grid>
                <Grid item direction="column" xs={4.5} style={{ marginTop: 50 }}>
                    <Box mt={22.5} p={2}>
                        <TextField
                            required
                            fullWidth
                            id="outlined-multiline-static"
                            label="Body"
                            multiline
                            rows={10}
                            value={content}
                            className={classes.textField}
                            InputLabelProps={{
                                style: { color: '#fff' },
                            }}
                            data-testid='Content'
                            onChange={(e) => setContent(e.target.value)}
                        />
                    </Box>
                    <Box m2={2} p={2}>
                        <Button id="submit-button" variant="contained" onClick={handleSubmit} style={{ backgroundColor: "#B18CFF" }} data-testid='Submit'>Submit</Button>
                    </Box>
                </Grid>

            </Grid >
        </div >
    )
}

const ThinkPieces = () => {

    return (
        <MuiThemeProvider theme={theme}>
            <div>
                <CssBaseline />
                <Paper>
                    <ThinkPiece />
                </Paper>
            </div>
        </MuiThemeProvider>
    );
}

ThinkPieces.propTypes = {
    classes: PropTypes.object.isRequired
};

export default ThinkPieces;