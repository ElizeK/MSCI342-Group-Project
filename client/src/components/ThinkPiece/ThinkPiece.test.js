import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import ThinkPiece from './index';


// test if thinkpiece renders on page 
describe('ThinkPiece', () => {

  test('renders component', () => {
    const { getByTestId } = render(<ThinkPiece />);
    const thinkPieceComponent = getByTestId('think-piece-component');
    expect(thinkPieceComponent).toBeInTheDocument();
  });

  it("checking title input", () => {
    const { getByTestId } = render(<ThinkPiece />);
    //expect(screen.getByText('Title')).toBeInTheDocument(); doesnt work - function not recognized after download 
    const titleInput = getByTestId('Title');
    expect(titleInput).toBeInTheDocument();
  })

  it("checking content input", () => {
    const { getByTestId } = render(<ThinkPiece />);
    const contentInput = getByTestId('Content');
    expect(contentInput).toBeInTheDocument();
  })

  it("checking URL input", () => {
    const { getByTestId } = render(<ThinkPiece />);
    const urlInput = getByTestId('URL');
    expect(urlInput).toBeInTheDocument();
  })

  it("checking Topic input", () => {
    const { getByTestId } = render(<ThinkPiece />);
    const topicInput = getByTestId('Topic');
    expect(topicInput).toBeInTheDocument();
  })


});



