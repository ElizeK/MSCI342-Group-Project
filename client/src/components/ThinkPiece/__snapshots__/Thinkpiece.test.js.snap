// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThinkPiece component shows error message on failed submission 1`] = `
<div>
  <Paper>
    <div>
      <div>
        <Toolbar>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginRight": 10,
              }
            }
          />
          <div
            className="MuiTypography-root makeStyles-navbarItem-12 MuiTypography-h6"
            sx={
              Object {
                "flexGrow": 1,
              }
            }
          >
            News App
          </div>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginRight": 50,
              }
            }
          />
          <Button
            color="inherit"
            onClick={[Function]}
          >
            <p
              className="MuiTypography-root makeStyles-navbarItem-12 MuiTypography-body1"
            >
              Home
            </p>
          </Button>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginRight": 50,
              }
            }
          />
          <Button
            color="inherit"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <p
              className="MuiTypography-root makeStyles-navbarItem-12 MuiTypography-body1"
            >
              Search
            </p>
          </Button>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginRight": 50,
              }
            }
          />
          <Button
            color="inherit"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <p
              className="MuiTypography-root makeStyles-navbarItem-12 MuiTypography-body1"
            >
              Thinkpiece
            </p>
          </Button>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginRight": 50,
              }
            }
          />
          <Button
            color="inherit"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <p
              className="MuiTypography-root makeStyles-navbarItem-12 MuiTypography-body1"
            >
              Profile
            </p>
          </Button>
        </Toolbar>
      </div>
      <Grid
        className="makeStyles-backgroundColor-2"
        container={true}
        direction="row"
        style={
          Object {
            "minHeight": "100vh",
          }
        }
      >
        <Grid
          item={true}
          style={
            Object {
              "marginTop": 50,
            }
          }
          xs={5}
        >
          <Box
            ml={7}
            p={2}
          >
            <p
              className="MuiTypography-root makeStyles-heading-10 MuiTypography-body1"
            >
              Think Piece
            </p>
            <p
              className="MuiTypography-root makeStyles-subHeading-11 MuiTypography-body1"
            >
              Write your own piece based on news you've read
            </p>
          </Box>
          <Box
            ml={7}
            p={2}
          >
            <TextField
              InputLabelProps={
                Object {
                  "style": Object {
                    "color": "#fff",
                  },
                }
              }
              className="makeStyles-textField-8"
              fullWidth={true}
              id="outlined-basic"
              label="Title"
              onChange={[Function]}
              required={true}
              value=""
              variant="outlined"
            />
          </Box>
          <Box
            ml={7}
            p={2}
          >
            <TextField
              InputLabelProps={
                Object {
                  "style": Object {
                    "color": "#fff",
                  },
                }
              }
              className="makeStyles-textField-8"
              fullWidth={true}
              id="outlined-multiline-required"
              label="Summary"
              maxRows={4}
              multiline={true}
              onChange={[Function]}
              required={true}
              value=""
            />
          </Box>
          <Box
            ml={7}
            p={2}
          >
            <FormControl
              fullWidth={true}
            >
              <InputLabel
                style={
                  Object {
                    "color": "#fff",
                  }
                }
              >
                Topic
              </InputLabel>
              <Select
                className="makeStyles-select-9"
                label="Topic"
                onChange={[Function]}
                required={true}
                style={
                  Object {
                    "color": "#fff",
                  }
                }
                value="General"
              >
                <MenuItem
                  value="Business"
                >
                  Business
                </MenuItem>
                <MenuItem
                  value="Entertainment"
                >
                  Entertainment
                </MenuItem>
                <MenuItem
                  value="General"
                >
                  General
                </MenuItem>
                <MenuItem
                  value="Health"
                >
                  Health
                </MenuItem>
                <MenuItem
                  value="Science"
                >
                  Science
                </MenuItem>
                <MenuItem
                  value="Sports"
                >
                  Sports
                </MenuItem>
                <MenuItem
                  value="Technology"
                >
                  Technology
                </MenuItem>
              </Select>
            </FormControl>
          </Box>
          <Box
            ml={7}
            p={2}
          >
            <TextField
              InputLabelProps={
                Object {
                  "style": Object {
                    "color": "#fff",
                  },
                }
              }
              className="makeStyles-textField-8"
              fullWidth={true}
              id="outlined-required"
              label="Article URL"
              onChange={[Function]}
              required={true}
              value=""
            />
          </Box>
        </Grid>
        <Grid
          item={true}
          xs={1}
        />
        <Grid
          direction="column"
          item={true}
          style={
            Object {
              "marginTop": 50,
            }
          }
          xs={4.5}
        >
          <Box
            mt={22.5}
            p={2}
          >
            <TextField
              InputLabelProps={
                Object {
                  "style": Object {
                    "color": "#fff",
                  },
                }
              }
              className="makeStyles-textField-8"
              fullWidth={true}
              id="outlined-multiline-static"
              label="Body"
              multiline={true}
              onChange={[Function]}
              required={true}
              rows={10}
              value=""
            />
          </Box>
          <Box
            m2={2}
            p={2}
          >
            <Button
              onClick={[Function]}
              style={
                Object {
                  "backgroundColor": "#B18CFF",
                }
              }
              variant="contained"
            >
              Submit
            </Button>
          </Box>
        </Grid>
      </Grid>
    </div>
  </Paper>
</div>
`;

exports[`ThinkPiece component shows error message on failed submission 2`] = `
<a
  className="hovered"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;

exports[`ThinkPiece component shows error message on failed submission 3`] = `
<a
  className="normal"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;
